### PR TITLE
Change c++ code to work with doubles

### PR DESCRIPTION
**Pull request by SlashScreen**
solves half of issue #30 - doesn't touch shader though

### CODE REVIEW

**Comment by SlashScreen** [comment]
Don't merge this until I've gone through everything

File: src/terrain_3d.cpp [hunk 1]
```diff
@@ -564,7 +564,7 @@ void Terrain3D::set_cast_shadows(GeometryInstance3D::ShadowCastingSetting p_shad
 	_update_instances();
 }
 
-void Terrain3D::set_cull_margin(float p_margin) {
+void Terrain3D::set_cull_margin(real_t p_margin) {
```
Comment (TokisanGames) [inline src/terrain_3d.cpp:567]: float for all usages of set/get cull_margin

File: src/terrain_3d.cpp [hunk 2]
```diff
@@ -603,10 +603,10 @@ void Terrain3D::snap(Vector3 p_cam_pos) {
 	int tile = 0;
 
 	for (int l = 0; l < _clipmap_levels; l++) {
-		float scale = float(1 << l);
+		real_t scale = real_t(1 << l);
 		Vector3 snapped_pos = (p_cam_pos / scale).floor() * scale;
-		Vector3 tile_size = Vector3(float(_clipmap_size << l), 0, float(_clipmap_size << l));
-		Vector3 base = snapped_pos - Vector3(float(_clipmap_size << (l + 1)), 0, float(_clipmap_size << (l + 1)));
+		Vector3 tile_size = Vector3(real_t(_clipmap_size << l), 0, real_t(_clipmap_size << l));
+		Vector3 base = snapped_pos - Vector3(real_t(_clipmap_size << (l + 1)), 0, real_t(_clipmap_size << (l + 1)));
```
Comment (TokisanGames) [inline src/terrain_3d.cpp:609]: All of this can be float. clipmap size will always be <1000

File: src/terrain_3d_storage.h [hunk 3]
```diff
@@ -20,7 +20,7 @@ class Terrain3DStorage : public Resource {
 
 	// Constants & Definitions
 
-	static inline const float CURRENT_VERSION = 0.84;
+	static inline const real_t CURRENT_VERSION = 0.84;
```
Comment (TokisanGames) [inline src/terrain_3d_storage.h:23]: float

File: src/terrain_3d_storage.h [hunk 4]
```diff
@@ -63,7 +63,7 @@ class Terrain3DStorage : public Resource {
 
 	// Storage Settings & flags
 
-	float _version = CURRENT_VERSION;
+	real_t _version = CURRENT_VERSION;
```
Comment (TokisanGames) [inline src/terrain_3d_storage.h:66]: float

**Comment by TokisanGames** [comment]
You know, thinking about it I think its fine that all instances here are switched to real_t. I was thinking using set_cull_margin(real_t) is a waste of space using a double when only a float is needed. But if we're doing it from gdscript what is coming in is a double anyway. 

The only concern I have is with saved variables. Say version for instance. If it's saved as a float into Storage. Then later read by a double engine, is it seamlessly converted? If it's saved as a double, can it be opened by a float engine without issue? If version and all of the other data can be read and written by both engines then it's fine to do everything as real_t.

Also include a line somewhere like __ready() that prints `LOG(ERROR, "Float: ", sizeof(float), " Real_t: ", sizeof(real_t));` and include the result here so we can make sure they are different.

**Comment by SlashScreen** [comment]
Asked AThousandShips: 

```
Me:
If a double (f64) value is serialized into a text file, and then opened using a single-precision compiled engine, what happens?
is it just truncated as a float, as if you were casting
And what about the reverse?

AThousandShips: 
Again most are stored as double, just some types rely on the precision setting
Like Vector2/3/4, Basis, Transform2D/3D, etc.
Most cases of float in the engine are f64, and serialisation isn't affected unless you're storing a type that depends
```
Shouldn't be an issue.

**Comment by SlashScreen** [comment]
```
SlugFiller:
Also, serialization into a text file truncates floating point way below the limit of single. There's an issue, and I believe, also a PR about it
```

File: src/terrain_3d.cpp [hunk 5]
```diff
@@ -75,7 +75,7 @@ void Terrain3D::_initialize() {
 
 void Terrain3D::__ready() {
 	_initialize();
-	LOG(WARN, "Float: ", sizeof(float), " Real_t: ", sizeof(real_t));
+	LOG(WARN, "Float: ", Variant(sizeof(float)), " Real_t: ", Variant(sizeof(real_t)));
```
Comment (TokisanGames) [inline src/terrain_3d.cpp:78]: We don't need these in the repo, I just want you to run it once to confirm that building with precision=double does something, and I'd like you to post your results in the comments.

Comment (SlashScreen) [reply src/terrain_3d.cpp:78]: Ah, Ok

**Comment by SlashScreen** [comment]
Got it to build, as I said in discord. Let me know if I need to change anything.

File: godot-cpp
Comment (TokisanGames) [inline godot-cpp:1]: Remove from PR. I will update godot-cpp later. Current commit tag is godot-4.1.1-stable

File: project/project.godot [hunk 6]
```diff
@@ -12,7 +12,7 @@ config_version=5
 
 config/name="Terrain3D"
 run/main_scene="res://demo/Demo.tscn"
-config/features=PackedStringArray("4.1")
+config/features=PackedStringArray("4.1", "Double Precision")
```
Comment (TokisanGames) [inline project/project.godot:15]: Remove from PR. Users can enable this

File: src/terrain_3d.cpp [hunk 7]
```diff
@@ -75,6 +75,7 @@ void Terrain3D::_initialize() {
 
 void Terrain3D::__ready() {
 	_initialize();
+	// LOG(WARN, "Float: ", Variant(int(sizeof(float))), " Real_t: ", Variant(int(sizeof(real_t))));
```
Comment (TokisanGames) [inline src/terrain_3d.cpp:78]: remove

File: src/util.cpp [hunk 8]
```diff
@@ -58,7 +58,7 @@ Vector2 Util::get_min_max(const Ref<Image> p_image) {
 }
 
 /**
- * Returns a Image of a float heightmap normalized to RGB8 greyscale and scaled
+ * Returns a Image of a real_t heightmap normalized to RGB8 greyscale and scaled
```
Comment (TokisanGames) [inline src/util.cpp:61]: **Returns a scaled, RGB8 greyscale Image of a heightmap normalized to its heights**

**Comment by SlashScreen** [comment]
oh dear. Will me removing godot-cpp from this PR cause issues in the main branch? The CI is freaking out right now.

File: src/terrain_3d_editor.cpp [hunk 9]
```diff
@@ -191,27 +191,27 @@ void Terrain3DEditor::_operate_map(Vector3 p_global_position, float p_camera_dir
 					storage->update_heights(destf);
 
 				} else if (map_type == Terrain3DStorage::TYPE_CONTROL) {
-					float alpha_clip = (brush_alpha < 0.1f) ? 0.0f : 1.0f;
-					int index_base = int(src.r * 255.0f);
-					int index_overlay = int(src.g * 255.0f);
+					real_t alpha_clip = (brush_alpha < real_t(0.1f)) ? 0.0f : 1.0f;
```
Comment (TokisanGames) [inline src/terrain_3d_editor.cpp:194]: Was there any sort of warning from the compiler comparing real_t brush_alpha and 0.1f?

If not, let's remove all of the literal conversions. I now think we only need the real_t conversions for ints. For floats to doubles it is free and unnecessary, unless there are warnings. Let's remove them all from here down to the last lerp on line 233.

https://stackoverflow.com/questions/13045485/division-of-double-with-a-float

The only need I see is where it is casting dest_index

Comment (SlashScreen) [reply src/terrain_3d_editor.cpp:194]: There were warnings when it came to function parameters; math.lerp didn't like an argument list of mixed floats and real_ts. I can remove the ones outside of those parameters, though.

**Comment by TokisanGames** [comment]
I looked closer at the godot source. Excluding modules & thirdparty code, .h/.cpp lines using float: 10212, real_t: 6091.

There are only 2 usages of float in all of the physics servers, none in navigation. 

Rendering is 99% float, needed only for locations or distance. The other uses are unnecessary.

The moviewriter is a new addition and has 3 floats, 0 real_t.

So my conclusions are:
1) it appears to be used where it's needed, so hopefully doubles actually does work.
2) Usage is all over the place and not very disciplined. When real_t is used in files, it is often used on inconsequential things and who knows if double support is completed and actually works.
3) Blanket search and replace will waste space for those using it, but unlikely to cause a problem.
4) Literals like  `5.0` are doubles, and `5.0f` are floats. Godot has `real_t x = 1.0f;` all over, so I don't think we need an explicit conversion in many cases unless the compiler produces a warning.

This link says float to double conversion on x87 fpu is free, and I've seen that other places as well.
https://stackoverflow.com/questions/13045485/division-of-double-with-a-float

So I think for consistency like the physics and navigation servers, we can use real_t everywhere. Let's wrap up this and merge it in. The shader portion can come separately.

Please make the following changes:
* Address all the individual comments I made
* Remove the changes to godot-cpp and project.godot. I will update the former soon along with a version upgrade. The godot-cpp tag is godot-4.1.1-stable
* Squash your changes somewhat or completely
* Rebase

File: src/terrain_3d_editor.cpp (hunk 9 above)
Comment (TokisanGames) [reply src/terrain_3d_editor.cpp:194]: Add any casts to remove warnings. Otherwise, we don't need it for float to double.

**Comment by SlashScreen** [comment]
I've addressed the individual comments. I will soon remove explicit conversions, squash (hopefully I don't screw it up this time), and rebase.

**Comment by TokisanGames** [comment]
> oh dear. Will me removing godot-cpp from this PR cause issues in the main branch? The CI is freaking out right now.

I don't mean *removing* godot-cpp from your repository, aka, deleting the folder and pushing that change. I mean reverting your godot-cpp to the previous version (`git checkout godot-4.1.1-stable`) and pushing that. That will revert the change you made and reset it back to the version in the main branch.

**Comment by TokisanGames** [comment]
Great job, thank you. In the future don't use `git merge`. You can do anything else like rebase, but let github do all merging.

**Comment by SlashScreen** [comment]
For the record, I never did git merge, I let GitHub do all of that. And, I never found any “merge” actions when I was rebasing.

**Pull request merged**